kong/tracing/instrumentation.lua

```diff
@@ -0,0 +1,229 @@
+  end
+end
+
+-- http_request (root span)
+function instrumentations.http_request()
+  local req = kong.request
+
+  local headers = req.get_headers()
+  local header_type, trace_id, span_id, parent_id, sampled, _ = propagation.parse(headers)
+  local method = req.get_method()
+  local path = req.get_path()
+  local span_name = method .. " " .. path
+
+  -- TODO(mayo): add host, port...
+  local active_span = instrument_tracer.active_span()
+  if not active_span then
+    active_span = instrument_tracer.start_span(span_name, {
+      trace_id = trace_id,
+      span_id = span_id,
+      parent_id = parent_id,
+      sampled = sampled,
+    })
+    instrument_tracer.set_active_span(active_span)
+  else
+    active_span.name = span_name
+    if trace_id then
+      active_span.trace_id = trace_id
+    end
+
+    if parent_id then
+      active_span.parent_id = parent_id
+    end
+
+    if sampled ~= nil then
+      active_span.sampled = sampled
+    end
+  end
+
+  active_span:set_attribute("http.host", req.get_host())
+
+  set_headers(header_type, active_span)
+end
+
+-- db query
+function instrumentations.db_query(connector)
+  local f = connector.query
+
+  local function wrap(self, sql, ...)
+    start_root_span()
+    local span = instrument_tracer.start_span("query", {
+      attributes = {
+        sql = sql,
+      }
+    })
+    local ret = pack(f(self, sql, ...))
+    span:finish()
+    return unpack(ret)
+  end
+
+  connector.query = wrap
+end
+
+-- router
+function instrumentations.router(router)
+  local f = router.exec
+  router.exec = wrap_func("router", f)
+end
+
+function instrumentations.runloop_log_before(ctx)
+  local root_span = instrument_tracer.active_span()
+  if root_span then
+    root_span:finish()
+  end
+end
+
+function instrumentations.runloop_log_after(ctx)
+  -- Clears the span table and put back the table pool,
+  -- this avoids reallocation.
+  -- The span table MUST NOT be used after released.
+  if type(ctx.KONG_SPANS) == "table" then
+    for _, span in ipairs(ctx.KONG_SPANS) do
+      tablepool.release("kong_span", span)
+    end
+    tablepool.release("kong_spans", ctx.KONG_SPANS)
+  end
+end
+
+function _M.init(config)
+  local trace_types = config.instrumentation_trace_types
+  local enabled = config.instrumentation_trace == true
+  local sampling_rate = config.instrumentation_trace_sampling_rate
+  if not enabled or type(trace_types) ~= "table" or sampling_rate <= 0 then
+    return
+  end
+
+  if trace_types[1] == "all" then
+    for typ, func in pairs(instrumentations) do
+      _M[typ] = func
+    end
+  end
+
+  for _, typ in ipairs(trace_types) do
+    local func = instrumentations[typ]
+    if func ~= nil then
+      _M[typ] = func
+    end
+  end
+
+  instrument_tracer = pdk_tracer.new("instrument", {
+    sampling_rate = sampling_rate,
+  })
+  instrument_tracer.set_global_tracer(instrument_tracer)
+end
+
+return setmetatable(_M, noop_mt)
```

Comment:
instead of using a metatable hack here, consider explictly set each function to a noop function, this can eliminate metatable lookup on hotpath. actually, instead of populating functions based on `trace_types` on Line210 to Line221, we can do the opposite: at first _M has all functions, and we noop those are not defined in `trace_types`.

Reply:
Thanks, sounds better!